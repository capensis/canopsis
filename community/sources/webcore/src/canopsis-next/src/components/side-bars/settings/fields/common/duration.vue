<template lang="pug">
  v-list-group
    v-list-tile(slot="activator")
      div(:class="validationHeaderClass")  {{ label }}
    v-container
      c-duration-field(
        v-field="value",
        :units-label="$t('common.unit')",
        :name="name",
        clearable
      )
</template>

<script>
import formMixin from '@/mixins/form';
import formValidationHeaderMixin from '@/mixins/form/validation-header';

export default {
  inject: ['$validator'],
  mixins: [formMixin, formValidationHeaderMixin],
  props: {
    value: {
      type: Object,
      required: false,
    },
    label: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: false,
    },
  },
};
</script>
