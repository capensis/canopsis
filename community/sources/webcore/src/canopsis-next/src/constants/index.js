export * from './recurrence-rule';
export * from './common';
export * from './entity';
export * from './alarm';
export * from './broadcast-message';
export * from './date';
export * from './dynamic-info';
export * from './event-filter';
export * from './pattern';
export * from './healthcheck';
export * from './idle-rule';
export * from './junit';
export * from './modal';
export * from './pbehavior';
export * from './permission';
export * from './remediation';
export * from './scenario';
export * from './snmp';
export * from './stats';
export * from './store';
export * from './widget';
export * from './kpi';
export * from './user';
export * from './service-weather';
export * from './map';
export * from './flowchart';
export * from './parameters';
export * from './declare-ticket';
export * from './link-rule';
export * from './external-data';
export * from './tag';
export * from './counter';
export * from './calendar';
export * from './icon';
export * from './availability';
