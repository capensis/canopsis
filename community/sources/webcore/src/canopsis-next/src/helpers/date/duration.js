import { isNil } from 'lodash';
import moment from 'moment';

import 'moment-duration-format';

import {
  AVAILABLE_SORTED_TIME_UNITS,
  DATETIME_FORMATS,
  DAYS_IN_MONTH,
  DAYS_IN_WEEK,
  DAYS_IN_YEAR,
  DEFAULT_DURATION_FORMAT,
  MONTHS_IN_YEAR,
  TIME_UNITS,
} from '@/constants';

/**
 * @typedef { "y" | "M" | "w" | "d" | "h" | "m" | "s" } DurationUnit
 */

/**
 * @typedef {Object} Duration
 * @property {number} seconds
 * @property {DurationUnit} unit
 */

/**
 * @typedef {Duration} DurationWithEnabled
 * @property {boolean} enabled
 */

/**
 * @typedef {Object} DurationForm
 * @property {number} value
 * @property {DurationUnit} unit
 */

/**
 * @typedef {DurationForm} DurationWithEnabledForm
 * @property {boolean} enabled
 */

/**
 * Convert duration from one unit to another unit
 *
 * @param {number} value
 * @param {DurationUnit} [fromUnit = TIME_UNITS.second]
 * @param {DurationUnit} [toUnit = TIME_UNITS.second]
 * @returns {number}
 */
export const convertUnit = (value, fromUnit = TIME_UNITS.second, toUnit = TIME_UNITS.second) => {
  if (fromUnit === toUnit) {
    return value;
  }

  /**
   * Using of the moment, we are faced with the problems of converting any units into months and years.
   * TODO: Should be removed after change format from seconds to value on backend size.
   * @link https://github.com/moment/moment/issues/5892
   */
  if (fromUnit === TIME_UNITS.year) {
    if (toUnit === TIME_UNITS.month) {
      return value * MONTHS_IN_YEAR;
    }

    if (toUnit === TIME_UNITS.week) {
      return value * (DAYS_IN_YEAR / DAYS_IN_WEEK);
    }

    return moment.duration(value * DAYS_IN_YEAR, TIME_UNITS.day).as(toUnit);
  }

  if (fromUnit === TIME_UNITS.month) {
    if (toUnit === TIME_UNITS.year) {
      return value / MONTHS_IN_YEAR;
    }

    return moment.duration(value * DAYS_IN_MONTH, TIME_UNITS.day).as(toUnit);
  }

  const momentDuration = moment.duration(value, fromUnit);

  if (toUnit === TIME_UNITS.month) {
    return momentDuration.as(TIME_UNITS.day) / DAYS_IN_MONTH;
  }

  if (toUnit === TIME_UNITS.year) {
    return momentDuration.as(TIME_UNITS.day) / DAYS_IN_YEAR;
  }

  return momentDuration.as(toUnit);
};

/**
 * Convert duration from unit to "seconds"
 *
 * @param {number} value
 * @param {DurationUnit} [unit = TIME_UNITS.second]
 * @returns {number}
 */
export const toSeconds = (value, unit = TIME_UNITS.second) => convertUnit(value, unit, TIME_UNITS.second);

/**
 * Convert duration from "seconds" to unit
 *
 * @param {number} value
 * @param {DurationUnit} [unit = TIME_UNITS.second]
 * @returns {number}
 */
export const fromSeconds = (value, unit = TIME_UNITS.second) => convertUnit(value, TIME_UNITS.second, unit);

/**
 * Convert Duration object to DurationForm
 *
 * @param {Duration} duration
 * @param {number} [duration.seconds = 1]
 * @param {DurationUnit} [duration.unit = TIME_UNITS.second]
 * @returns {DurationForm}
 */
export const durationToForm = ({ seconds = 1, unit = TIME_UNITS.second } = {}) => ({
  unit,
  value: fromSeconds(seconds, unit),
});

/**
 * Convert DurationForm object to Duration
 *
 * @param {DurationForm} duration
 * @param {number} [duration.value = 0]
 * @param {DurationUnit} [duration.unit = TIME_UNITS.second]
 * @returns {Duration}
 */
export const formToDuration = ({ value = 0, unit = TIME_UNITS.second } = {}) => ({
  unit,
  seconds: toSeconds(value, unit),
});

/**
 * Convert DurationWithEnabled object to DurationWithEnabledForm
 *
 * @param {DurationWithEnabled} duration
 * @return {DurationWithEnabledForm}
 */
export const durationWithEnabledToForm = ({ seconds, unit, enabled = false } = {}) => ({
  ...durationToForm({ seconds, unit }),

  enabled,
});

/**
 * Convert DurationWithEnabled object to DurationWithEnabledForm
 *
 * @param {DurationWithEnabledForm} duration
 * @return {DurationWithEnabled}
 */
export const formToDurationWithEnabled = ({ value, unit, enabled }) => ({
  ...formToDuration({ value, unit }),

  enabled,
});

/**
 * Get max available interval value
 *
 * @param {DurationForm} [durationForm = { value: 0, unit: TIME_UNITS.second }]
 * @param {DurationUnit[]} [availableUnits = AVAILABLE_SORTED_TIME_UNITS]
 * @return {DurationForm}
 */
export const formToMaxByAvailableUnitsForm = (
  durationForm = { value: 0, unit: TIME_UNITS.second },
  availableUnits = AVAILABLE_SORTED_TIME_UNITS,
) => {
  const { value, unit } = durationForm;
  let unitValue = value;

  const maxUnit = availableUnits.find((availableUnit) => {
    unitValue = Math.floor(convertUnit(value, unit, availableUnit));
    return unitValue;
  });

  return {
    value: unitValue,
    unit: maxUnit || unit,
  };
};

/**
 * Convert duration to more readable format
 *
 * @param {number | Duration | DurationForm} duration
 * @param {string} [format = DEFAULT_DURATION_FORMAT]
 * @param {DurationUnit} [unit = TIME_UNITS.second]
 * @returns {string}
 */
export const durationToString = (duration, format = DEFAULT_DURATION_FORMAT, unit = TIME_UNITS.second) => {
  if (isNil(duration)) {
    return '';
  }

  /**
   * TODO: Should be removed after duration refactoring
   */
  const durationValue = duration ? (duration.seconds || duration.value || duration) : duration;
  const resultFormat = DATETIME_FORMATS[format] || format;

  return moment.duration(durationValue, unit).format(resultFormat, { trim: 'both final' }) || '0s';
};
