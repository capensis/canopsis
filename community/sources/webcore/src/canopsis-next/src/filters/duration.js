import moment from 'moment';
import { isNil } from 'lodash';

import 'moment-duration-format';

import { DATETIME_FORMATS, DEFAULT_DURATION_FORMAT, TIME_UNITS } from '@/constants';

/**
 * Duration filter
 *
 * @param {number | Duration | DurationForm} duration
 * @param {string} [format = DEFAULT_DURATION_FORMAT]
 * @param {DurationUnit} [unit = TIME_UNITS.second]
 * @returns {string}
 */
export default function (duration, format = DEFAULT_DURATION_FORMAT, unit = TIME_UNITS.second) {
  if (isNil(duration)) {
    return '';
  }

  /**
   * TODO: Should be removed after duration refactoring
   */
  const durationValue = duration ? (duration.seconds || duration.value || duration) : duration;
  const resultFormat = DATETIME_FORMATS[format] || format;

  return moment.duration(durationValue, unit).format(resultFormat, { trim: 'both final' }) || '0s';
}
