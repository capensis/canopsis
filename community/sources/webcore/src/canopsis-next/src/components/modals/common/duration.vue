<template>
  <v-form @submit.prevent="submit">
    <modal-wrapper close>
      <template #title="">
        <span>{{ title }}</span>
      </template>
      <template #text="">
        <c-duration-field
          v-model="form"
          :label="label"
          :units="units"
          required
        />
      </template>
      <template #actions="">
        <v-btn
          depressed
          text
          @click="close"
        >
          {{ $t('common.cancel') }}
        </v-btn>
        <v-btn
          :disabled="isDisabled"
          :loading="submitting"
          class="primary"
          type="submit"
        >
          {{ $t('common.submit') }}
        </v-btn>
      </template>
    </modal-wrapper>
  </v-form>
</template>

<script>
import { computed, ref } from 'vue';

import { MODALS, VALIDATION_DELAY } from '@/constants';

import { durationToForm } from '@/helpers/date/duration';

import { useI18n } from '@/hooks/i18n';
import { useInnerModal } from '@/hooks/modals';
import { useSubmittableForm } from '@/hooks/submittable-form';
import { useFormConfirmableCloseModal } from '@/hooks/confirmable-modal';

import ModalWrapper from '../modal-wrapper.vue';

export default {
  name: MODALS.duration,
  $_veeValidate: {
    validator: 'new',
    delay: VALIDATION_DELAY,
  },
  components: { ModalWrapper },
  props: {
    modal: {
      type: Object,
      required: true,
    },
  },
  setup(props) {
    const { t, tc } = useI18n();
    const { config, close } = useInnerModal(props);

    const form = ref(durationToForm(config.value.duration));

    const title = computed(() => config.value.title ?? t('common.duration'));
    const label = computed(() => config.value.label ?? t('common.duration'));
    const units = computed(() => (
      config.value.units?.map(unit => ({ ...unit, text: tc(unit.text, form.value.value) }))
    ));

    const { submit, isDisabled, submitting } = useSubmittableForm({
      form,
      method: async () => {
        await config.value.action?.(form);
        close();
      },
    });

    useFormConfirmableCloseModal({ form, submit, close });

    return {
      form,
      title,
      label,
      units,
      isDisabled,
      submitting,

      close,
      submit,
    };
  },
};
</script>
