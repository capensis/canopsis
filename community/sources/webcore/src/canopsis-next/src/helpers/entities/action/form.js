import { omit } from 'lodash';

import { ENTITIES_STATES, ACTION_TYPES, PATTERNS_FIELDS, OLD_PATTERNS_FIELDS } from '@/constants';

import { uid } from '@/helpers/uid';
import { durationToForm } from '@/helpers/date/duration';
import { getLocaleTimezone } from '@/helpers/date/date';

import { formToPbehavior, pbehaviorToForm, pbehaviorToRequest } from '../pbehavior/form';
import { requestToForm, formToRequest } from '../shared/request/form';
import { eventToAssociateTicketForm, formToAssociateTicketEvent } from '../associate-ticket/event/form';
import {
  declareTicketRuleWebhookDeclareTicketToForm,
  formToDeclareTicketRuleWebhookDeclareTicket,
} from '../declare-ticket/rule/form';
import { filterPatternsToForm, formFilterToPatterns } from '../filter/form';

/**
 * @typedef {
 *   "snooze" |
 *   "pbehavior" |
 *   "changestate" |
 *   "webhook" |
 *   "ack" |
 *   "ackremove" |
 *   "assocticket" |
 *   "cancel"
 * } ActionType
 */

/**
 * @typedef {Object} ActionForwardAuthorParameters
 * @property {boolean} [forward_author]
 * @property {string} [author]
 */

/**
 * @typedef {ActionForwardAuthorParameters} ActionDefaultParameters
 * @property {string} output
 */

/**
 * @typedef {ActionDefaultParameters} ActionSnoozeParameters
 * @property {Duration} duration
 */

/**
 * @typedef {ActionDefaultParameters} ActionChangeStateParameters
 * @property {number} state
 */

/**
 * @typedef {ActionDefaultParameters & AssociateTicketEvent} ActionAssocTicketParameters
 */

/**
 * @typedef {Object} ActionWebhookParameters
 * @property {Request} request
 * @property {?DeclareTicketRuleWebhookDeclareTicket} [declare_ticket]
 * @property {boolean} [forward_author]
 * @property {boolean} skip_for_child
 * @property {boolean} skip_for_instruction
 * @property {string} [author]
 */

/**
 * @typedef {ActionWebhookParameters} ActionWebhookFormParameters
 * @property {RequestForm} request
 * @property {DeclareTicketRuleWebhookDeclareTicketForm} declare_ticket
 */

/**
 * @typedef {
 *   Pbehavior |
 *   ActionDefaultParameters |
 *   ActionSnoozeParameters |
 *   ActionChangeStateParameters |
 *   ActionWebhookParameters |
 *   ActionAssocTicketParameters
 * } ActionParameters
 */

/**
 * @typedef {FilterPatterns} Action
 * @property {ActionType} type
 * @property {boolean} drop_scenario_if_not_matched
 * @property {boolean} emit_trigger
 * @property {string} comment
 * @property {ActionParameters} parameters
 */

/**
 * @typedef {
 *   PbehaviorForm |
 *   ActionDefaultParameters |
 *   ActionSnoozeParameters |
 *   ActionChangeStateParameters |
 *   ActionWebhookFormParameters |
 *   ActionAssocTicketParameters
 * } ActionFormParameters
 */

/**
 * @typedef {Action} ActionForm
 * @property {FilterPatternsForm} patterns
 * @property {Object.<ActionType, ActionFormParameters>} parameters
 */

/**
 * Check action type is pbehavior
 *
 * @param {ActionType} type
 * @return {boolean}
 */
export const isPbehaviorActionType = type => type === ACTION_TYPES.pbehavior;

/**
 * Check action type is webhook
 *
 * @param {ActionType} type
 * @return {boolean}
 */
export const isWebhookActionType = type => type === ACTION_TYPES.webhook;

/**
 * Check if action type is associate ticket
 *
 * @param {ActionType} type
 * @returns {boolean}
 */
export const isAssociateTicketActionType = type => type === ACTION_TYPES.assocticket;

/**
 * Convert action parameters to form
 *
 * @param {ActionForwardAuthorParameters | {}} [parameters = {}]
 * @returns {ActionForwardAuthorParameters}
 */
const defaultActionForwardAuthorToForm = (parameters = {}) => ({
  forward_author: parameters.forward_author ?? true,
  author: parameters.author ?? '',
});

/**
 * Convert action parameters to form
 *
 * @param {ActionDefaultParameters | {}} [parameters = {}]
 * @returns {ActionDefaultParameters}
 */
const defaultActionParametersToForm = (parameters = {}) => ({
  ...defaultActionForwardAuthorToForm(parameters),
  output: parameters.output ?? '',
});

/**
 * Convert action webhook parameters to form
 *
 * @param {ActionWebhookParameters} [parameters = {}]
 * @returns {ActionWebhookFormParameters}
 */
const webhookActionParametersToForm = (parameters = {}) => ({
  ...defaultActionForwardAuthorToForm(parameters),
  declare_ticket: declareTicketRuleWebhookDeclareTicketToForm(parameters.declare_ticket),
  request: requestToForm(parameters.request),
  skip_for_child: parameters.skip_for_child ?? false,
  skip_for_instruction: parameters.skip_for_instruction ?? false,
});

/**
 * Convert action snooze parameters to form
 *
 * @param {ActionSnoozeParameters | {}} [parameters = {}]
 * @returns {ActionSnoozeParameters}
 */
const snoozeActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  duration: durationToForm(parameters.duration),
});

/**
 * Convert action snooze parameters to form
 *
 * @param {ActionChangeStateParameters | {}} [parameters = {}]
 * @returns {ActionChangeStateParameters}
 */
const changeStateActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  state: parameters.state ?? ENTITIES_STATES.minor,
});

/**
 * Convert action assoc ticket parameters to form
 *
 * @param {ActionAssocTicketParameters | {}} [parameters = {}]
 * @returns {ActionAssocTicketParameters}
 */
const assocTicketActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  ...omit(eventToAssociateTicketForm(parameters), ['ticket_comment']),
});

/**
 * Convert action pbehavior parameters to form
 *
 * @param {Pbehavior} [parameters = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @returns {PbehaviorForm}
 */
const pbehaviorActionParametersToForm = (parameters = {}, timezone = getLocaleTimezone()) => {
  const pbehaviorForm = pbehaviorToForm(parameters, null, timezone);

  pbehaviorForm.start_on_trigger = !!parameters.start_on_trigger;
  pbehaviorForm.duration = durationToForm(parameters.duration);

  return pbehaviorForm;
};

/**
 * Prepare parameters for all action types
 *
 * @returns {Object.<ActionType, ActionFormParameters>}
 */
const prepareDefaultActionParameters = () => ({
  [ACTION_TYPES.snooze]: snoozeActionParametersToForm(),
  [ACTION_TYPES.pbehavior]: pbehaviorActionParametersToForm(),
  [ACTION_TYPES.changeState]: changeStateActionParametersToForm(),
  [ACTION_TYPES.ack]: defaultActionParametersToForm(),
  [ACTION_TYPES.ackremove]: defaultActionParametersToForm(),
  [ACTION_TYPES.assocticket]: assocTicketActionParametersToForm(),
  [ACTION_TYPES.cancel]: defaultActionParametersToForm(),
  [ACTION_TYPES.webhook]: webhookActionParametersToForm(),
});

/**
 * Convert action parameters to form
 *
 * @param {Action} action
 * @param {string} timezone
 * @returns {Object.<ActionType, ActionParameters>}
 */
export const actionParametersToForm = (action, timezone) => {
  const parameters = prepareDefaultActionParameters();

  if (!action.type || !action.parameters) {
    return parameters;
  }

  const parametersPreparers = {
    [ACTION_TYPES.snooze]: snoozeActionParametersToForm,
    [ACTION_TYPES.webhook]: webhookActionParametersToForm,
    [ACTION_TYPES.pbehavior]: pbehaviorActionParametersToForm,
    [ACTION_TYPES.assocticket]: assocTicketActionParametersToForm,
    [ACTION_TYPES.changeState]: changeStateActionParametersToForm,
  };

  const prepareParametersToFormFunction = parametersPreparers[action.type];

  parameters[action.type] = prepareParametersToFormFunction
    ? prepareParametersToFormFunction(action.parameters, timezone)
    : defaultActionParametersToForm({ ...action.parameters });

  return parameters;
};

/**
 * Convert action to form
 *
 * @param {Action} [action = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @returns {ActionForm}
 */
export const actionToForm = (action = {}, timezone = getLocaleTimezone()) => ({
  type: action.type ?? ACTION_TYPES.snooze,
  key: uid(),
  parameters: actionParametersToForm(action, timezone),
  drop_scenario_if_not_matched: !!action.drop_scenario_if_not_matched,
  emit_trigger: !!action.emit_trigger,
  comment: action.comment ?? '',
  patterns: filterPatternsToForm(
    action,
    [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity],
    [OLD_PATTERNS_FIELDS.alarm, OLD_PATTERNS_FIELDS.entity],
  ),
});

/**
 * Convert pbehavior parameters to action
 *
 * @param {ActionWebhookFormParameters | {}} [parameters = {}]
 * @return {ActionWebhookParameters}
 */
export const formToWebhookActionParameters = (parameters = {}) => ({
  declare_ticket: formToDeclareTicketRuleWebhookDeclareTicket(parameters.declare_ticket),
  request: formToRequest(parameters.request),
  skip_for_child: parameters.skip_for_child,
  skip_for_instruction: parameters.skip_for_instruction,
});

/**
 * Convert pbehavior parameters to action
 *
 * @param {PbehaviorForm | {}} [parameters = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @return {PbehaviorRequest}
 */
export const formToPbehaviorActionParameters = (parameters = {}, timezone = getLocaleTimezone()) => {
  const pbehavior = formToPbehavior(omit(parameters, ['start_on_trigger', 'duration']), timezone);

  if (parameters.start_on_trigger) {
    pbehavior.start_on_trigger = parameters.start_on_trigger;
    pbehavior.duration = parameters.duration;
  }

  return pbehaviorToRequest(pbehavior);
};

/**
 * Convert form to action parameters
 *
 * @param {ActionForm} form
 * @param {string} [timezone]
 * @returns {ActionParameters}
 */
const formToActionParameters = (form, timezone) => {
  const {
    forward_author: forwardAuthor,
    author,
    ...parametersByCurrentType
  } = form.parameters[form.type];

  const parametersPreparers = {
    [ACTION_TYPES.webhook]: formToWebhookActionParameters,
    [ACTION_TYPES.pbehavior]: formToPbehaviorActionParameters,
    [ACTION_TYPES.assocticket]: formToAssociateTicketEvent,
  };

  const prepareParametersToAction = parametersPreparers[form.type];
  const parameters = prepareParametersToAction
    ? prepareParametersToAction(parametersByCurrentType, timezone)
    : parametersByCurrentType;

  if (!isPbehaviorActionType(form.type)) {
    parameters.forward_author = forwardAuthor;

    if (!forwardAuthor) {
      parameters.author = author;
    }
  }

  return parameters;
};

/**
 * Convert form to action
 *
 * @param {ActionForm} form
 * @param {string} [timezone]
 * @returns {Action}
 */
export const formToAction = (form, timezone) => ({
  ...omit(form, ['key', 'patterns']),
  ...formFilterToPatterns(form.patterns, [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity]),
  parameters: formToActionParameters(form, timezone),
});
