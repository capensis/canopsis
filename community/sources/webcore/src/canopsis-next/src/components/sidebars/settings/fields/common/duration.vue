<template lang="pug">
  v-list-group
    v-list-tile(slot="activator")
      div(:class="validationHeaderClass")  {{ label }}
    v-container
      c-duration-field(
        v-field="value",
        :units-label="$t('common.unit')",
        :name="name",
        clearable
      )
</template>

<script>
import { formMixin, formValidationHeaderMixin } from '@/mixins/form';

export default {
  inject: ['$validator'],
  mixins: [formMixin, formValidationHeaderMixin],
  props: {
    value: {
      type: Object,
      required: false,
    },
    label: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: false,
    },
  },
};
</script>
