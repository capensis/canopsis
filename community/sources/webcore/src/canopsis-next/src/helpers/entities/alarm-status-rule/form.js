import { omit } from 'lodash';

import { PATTERNS_FIELDS, TIME_UNITS } from '@/constants';

import { durationToForm } from '@/helpers/date/duration';
import { filterPatternsToForm, formFilterToPatterns } from '@/helpers/entities/filter/form';

/**
 * @typedef {FilterPatterns} AlarmStatusRule
 * @property {string} name
 * @property {string} description
 * @property {Duration} duration
 * @property {number} priority
 * @property {number} [freq_limit]
 */

/**
 * @typedef {Object} AlarmStatusRuleForm
 * @property {string} name
 * @property {string} description
 * @property {Duration} duration
 * @property {number} priority
 * @property {FilterPatternsForm} patterns
 */

/**
 * Convert alarm status rule object to form compatible object
 *
 * @param {AlarmStatusRule | {}} [rule = {}]
 * @param {boolean} [flapping = false]
 * @return {AlarmStatusRuleForm}
 */
export const alarmStatusRuleToForm = (rule = {}, flapping = false) => {
  const form = {
    name: rule.name ?? '',
    duration: rule.duration
      ? durationToForm(rule.duration)
      : { value: 1, unit: TIME_UNITS.minute },
    priority: rule.priority,
    description: rule.description ?? '',
    patterns: filterPatternsToForm(rule, [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity]),
  };

  if (flapping) {
    form.freq_limit = rule.freq_limit ?? 1;
  }

  return form;
};

/**
 * Convert form compatible object to alarm status rule object
 *
 * @param {AlarmStatusRuleForm} form
 * @return {AlarmStatusRule}
 */
export const formToAlarmStatusRule = form => ({
  ...omit(form, ['patterns']),
  ...formFilterToPatterns(form.patterns),
});
