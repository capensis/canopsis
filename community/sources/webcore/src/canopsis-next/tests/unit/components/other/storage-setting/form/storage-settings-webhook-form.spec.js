import { generateRenderer } from '@unit/utils/vue';
import { createCheckboxInputStub } from '@unit/stubs/input';
import { randomDurationEnabledValue } from '@unit/utils/duration';

import { TIME_UNITS } from '@/constants';

import { dataStorageSettingsToForm } from '@/helpers/entities/data-storage/form';

import CInformationBlock from '@/components/common/block/c-information-block.vue';
import StorageSettingsWebhookForm from '@/components/other/storage-setting/form/storage-settings-webhook-form.vue';

const snapshotStubs = {
  'c-information-block': CInformationBlock,
  'c-help-icon': true,
  'c-enabled-duration-field': true,
  'storage-settings-history-message': true,
  'c-enabled-field': true,
};
const stubs = {
  ...snapshotStubs,
  'v-checkbox': createCheckboxInputStub('v-checkbox'),
};

const selectWebhookDeleteAfterField = wrapper => wrapper.find('c-enabled-duration-field-stub');

describe('storage-settings-webhook-form', () => {
  const form = {
    log_credentials: true,
    delete_after: {
      value: 60,
      unit: TIME_UNITS.day,
      enabled: true,
    },
  };

  const factory = generateRenderer(StorageSettingsWebhookForm, { stubs });
  const snapshotFactory = generateRenderer(StorageSettingsWebhookForm, {
    stubs: snapshotStubs,
  });

  test('Webhook delete after changed after trigger enabled duration field', () => {
    const wrapper = factory({
      propsData: {
        form,
      },
    });

    const newValue = randomDurationEnabledValue();

    selectWebhookDeleteAfterField(wrapper).triggerCustomEvent('input', newValue);

    expect(wrapper).toEmitInput({ ...form, delete_after: newValue });
  });

  test('Renders `storage-settings-webhook-form` with default form', () => {
    const wrapper = snapshotFactory({
      propsData: {
        form: dataStorageSettingsToForm().webhook,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  test('Renders `storage-settings-webhook-form` with custom form and history', () => {
    const wrapper = snapshotFactory({
      propsData: {
        form,
        history: 1611250000,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
