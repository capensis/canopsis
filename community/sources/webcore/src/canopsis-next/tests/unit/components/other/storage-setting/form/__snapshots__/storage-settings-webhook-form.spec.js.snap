// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`storage-settings-webhook-form Renders \`storage-settings-webhook-form\` with custom form and history 1`] = `
<div
  class="layout column"
>
  <div
    class="layout c-information-block column"
  >
    <div
      class="layout"
    >
      <div
        class="text-subtitle-1 font-weight-bold"
      >
        
      Webhooks data storage
    
      </div>
       
      <c-help-icon-stub
        color="info"
        icon-class="ml-2"
        max-width="300"
        text="All webhook requests history is kept in logs"
        top=""
      />
    </div>
     
    <div
      class="layout mt-2"
    >
      <span
        class="text--secondary"
      >
        <storage-settings-history-message-stub
          history="1611250000"
        />
      </span>
    </div>
      
    <c-enabled-duration-field-stub
      help-text="All webhook requests history older than the defined time period will be deleted"
      label="Clear webhooks logs after"
      name="webhook.delete_after"
      value="[object Object]"
    />
     
    <c-enabled-field-stub
      name="webhook.log_credentials"
      value="true"
    />
  </div>
</div>
`;

exports[`storage-settings-webhook-form Renders \`storage-settings-webhook-form\` with default form 1`] = `
<div
  class="layout column"
>
  <div
    class="layout c-information-block column"
  >
    <div
      class="layout"
    >
      <div
        class="text-subtitle-1 font-weight-bold"
      >
        
      Webhooks data storage
    
      </div>
       
      <c-help-icon-stub
        color="info"
        icon-class="ml-2"
        max-width="300"
        text="All webhook requests history is kept in logs"
        top=""
      />
    </div>
     
    <!---->
      
    <c-enabled-duration-field-stub
      help-text="All webhook requests history older than the defined time period will be deleted"
      label="Clear webhooks logs after"
      name="webhook.delete_after"
      value="[object Object]"
    />
     
    <c-enabled-field-stub
      name="webhook.log_credentials"
    />
  </div>
</div>
`;
